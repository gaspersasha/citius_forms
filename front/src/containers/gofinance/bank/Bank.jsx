import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import { Button } from '~components';
import { INPUT_STATUS, VALIDATION_TYPES } from '~constants';
import {
  checkNotEmpty,
  stringLengthIsValid,
  prepare,
  validate,
  simplify,
  isValid,
  findMaxSteps,
  makeYearOptions,
  scrollTo,
} from '~utils';
import { handleGTM } from '../helpers';
import { AccountInfoStep, YearsAtBankStep } from '.';
import s from './bank.module.sass';
import financeStyles from '../styles/finance.module.sass';

const createDatesForBankList = (
  startDateOptionValue,
  endDateOptionsValue,
  dateTypeName
) =>
  Array(endDateOptionsValue - startDateOptionValue + 1)
    .fill()
    .map((item, index) => {
      const currentDateOptionValue = startDateOptionValue + index;

      return {
        option: `${currentDateOptionValue} ${dateTypeName}`,
        value: `${currentDateOptionValue}`,
      };
    });

const nextButtonIDs = {
  bankDetails1: 'ssFinanceBankDetails1Submit',
  bankDetails2: 'ssFinanceBankDetails2Submit',
};

const Bank = ({ bankDetails, onSubmit, step, isDataSending, leadgen }) => {
  const initialState = {
    accountNumber: {
      value: bankDetails.accountNumber || '',
      status: INPUT_STATUS.DEFAULT,
      validationType: VALIDATION_TYPES.ACCOUNT_NUMBER,
      validLength: 8,
      step: 0,
    },
    sortCode: {
      value: bankDetails.sortCode || '',
      status: INPUT_STATUS.DEFAULT,
      validationType: VALIDATION_TYPES.SORTCODE,
      validLength: 8,
      step: 0,
    },
    yearsAtBank: {
      value: bankDetails.yearsAtBank || '0',
      status: INPUT_STATUS.DEFAULT,
      validationType: VALIDATION_TYPES.NONE,
      step: 1,
    },
    monthAtBank: {
      value: bankDetails.monthAtBank || '0',
      status: INPUT_STATUS.DEFAULT,
      validationType: VALIDATION_TYPES.NONE,
      step: 1,
    },
    updatedAtStep: 0,
  };

  const [errors, setErrors] = useState({});
  const [state, setState] = useState(initialState);
  const maxSteps = findMaxSteps(state);
  const isDataSendingInProgress = isDataSending && step === 1;

  useEffect(() => {
    state.updatedAtStep && onSubmit(simplify(state), maxSteps);
  }, [state.updatedAtStep]);

  const dateValues = {
    yearsAtBankList: makeYearOptions(0, 4),
    monthAtBankList: createDatesForBankList(0, 11, 'Month(s)'),
  };

  const handleTimeAtBankChange = ({ target: { value, name } }) => {
    setState((prev) => ({
      ...prev,
      [name]: {
        ...prev[name],
        value,
      },
    }));
  };

  const handleAccountInfoChange = ({ target: { value, name } }) => {
    if (value.length > state[name]?.validLength) return;
    setState((prev) => ({
      ...prev,
      [name]: {
        ...prev[name],
        value,
      },
    }));
  };

  const handleBlurAccountInfoChange = ({ target: { value, name } }) => {
    const status = checkNotEmpty(value)
      ? INPUT_STATUS.INVALID
      : INPUT_STATUS.DEFAULT;

    setState((prev) => ({
      ...prev,
      [name]: {
        ...prev[name],
        value,
        status: stringLengthIsValid(value, prev[name]?.validLength)
          ? INPUT_STATUS.VALID
          : status,
      },
    }));
  };

  const addErrorTo = (key, newError) =>
    setErrors((prevErrors) => ({
      ...prevErrors,
      [key]: newError,
    }));

  const removeErrorForm = (key) =>
    setErrors((prevErrors) => ({
      ...prevErrors,
      [key]: null, // or undefined
    }));

  const handleSubmit = (e) => {
    e.preventDefault();
    e.stopPropagation();

    if (leadgen) {
      const accountNumber = state.accountNumber.value.length === 8;
      const sortCode = state.sortCode.value.length === 8;

      if (!accountNumber) {
        addErrorTo('accountNumber', 'invalid');
      } else {
        removeErrorForm('accountNumber');
      }

      if (!sortCode) {
        addErrorTo('sortCode', 'invalid');
      } else {
        removeErrorForm('sortCode');
      }

      if (accountNumber && sortCode) {
        if (isDataSendingInProgress) return;

        const updated = validate(prepare(state, step));
        const tryAgain = !isValid(updated);

        if (tryAgain) return scrollTo('.error-input-label');
        handleGTM(updated, ['accountNumber', 'yearsAtBank']);

        setState({
          ...state,
          ...updated,
          updatedAtStep: state.updatedAtStep + 1,
        });
      }

      return;
    }

    if (isDataSendingInProgress) return;

    const updated = validate(prepare(state, step));
    const tryAgain = !isValid(updated);

    if (tryAgain) return scrollTo('.error-input-label');
    handleGTM(updated, ['accountNumber', 'yearsAtBank']);

    setState({
      ...state,
      ...updated,
      updatedAtStep: state.updatedAtStep + 1,
    });
  };

  const isMonthDisabled = () => {
    const length = dateValues.yearsAtBankList.length - 1;

    return state.yearsAtBank.value === dateValues.yearsAtBankList[length].value;
  };

  const getNextButtonID = () => {
    switch (step) {
      case 0:
        return nextButtonIDs.bankDetails1;
      case 1:
        return nextButtonIDs.bankDetails2;
      default:
        return null;
    }
  };

  const buttonClasses = cn(
    { 'button-forward-loading': isDataSendingInProgress },
    `${financeStyles.confirm}`
  );

  return (
    <div className={s.formSectionWrapper}>
      <p
        className={cn(financeStyles.formPreamble, s.bankPreamble)}
        data-bind="bankDetailform"
      >
        Some <b>lenders require basic bank details</b>. Filling in the details
        below will get you a wider range of quotes
      </p>

      {step === 0 && (
        <AccountInfoStep
          onChange={handleAccountInfoChange}
          onBlur={handleBlurAccountInfoChange}
          accountNumber={state.accountNumber}
          sortCode={state.sortCode}
          errors={errors}
          removeErrorForm={removeErrorForm}
        />
      )}

      {step === 1 && (
        <YearsAtBankStep
          onChange={handleTimeAtBankChange}
          yearsAtBank={state.yearsAtBank}
          yearsAtBankOptions={dateValues.yearsAtBankList}
          monthAtBank={state.monthAtBank}
          monthAtBankOptions={dateValues.monthAtBankList}
          monthAtBankIsDisabled={isMonthDisabled()}
        />
      )}

      <Button
        onClick={handleSubmit}
        type="submit"
        className={buttonClasses}
        id={getNextButtonID()}
      >
        {isDataSendingInProgress ? 'Loading...' : 'Next'}
      </Button>
    </div>
  );
};

Bank.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  step: PropTypes.number.isRequired,
  isDataSending: PropTypes.bool.isRequired,
  bankDetails: PropTypes.shape({
    accountNumber: PropTypes.string,
    sortCode: PropTypes.string,
    yearsAtBank: PropTypes.string,
    monthAtBank: PropTypes.string,
  }),
  leadgen: PropTypes.bool.isRequired,
};

Bank.defaultProps = {
  bankDetails: {
    accountNumber: '',
    sortCode: '',
    yearsAtBank: '0',
    monthAtBank: '0',
  },
};

export default Bank;
