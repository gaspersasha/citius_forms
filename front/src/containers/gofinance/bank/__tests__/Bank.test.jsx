import React from 'react';
import Bank from '../Bank';

mockUseUserContext({
  user: { id: -1 },
});

describe('<Bank />', () => {
  const props = {
    onSubmit: jest.fn(),
    step: 0,
    isLoading: false,
    isDataSending: false,
  };

  const wrapper = mount(<Bank {...props} />);
  const AN_SELECTOR = 'input[type="number"][name="accountNumber"]';
  const SC_SELECTOR = 'input[type="tel"][name="sortCode"]';
  const SUBMIT = 'button';

  it('renders without errors', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('invalid form can`t be submitted', () => {
    const numberEvent = { target: { value: '123', name: 'accountNumber' } };

    wrapper.find(AN_SELECTOR).simulate('change', numberEvent);
    wrapper.find(SUBMIT).simulate('click');
    expect(wrapper.props().onSubmit).toHaveBeenCalledTimes(0);
  });

  it('account number field responds to valid changes', () => {
    const $el = wrapper.find(AN_SELECTOR);
    const event = { target: { value: '123', name: 'accountNumber' } };

    $el.simulate('change', event);
    expect($el.props().value).toBe('123');
  });

  it('method handleBlurAccountInfoChange should work correct', () => {
    const $an = wrapper.find(AN_SELECTOR);
    const $sc = wrapper.find(SC_SELECTOR);
    const AccountNumberEvent = {
      target: { value: '123', name: 'accountNumber' },
    };
    const sortCodetEvent = { target: { value: '12-34-56', name: 'sortCode' } };

    $an.simulate('change', AccountNumberEvent);
    $an.simulate('blur');
    $sc.simulate('change', sortCodetEvent);
    $sc.simulate('blur');
    expect($an.hasClass('error-input-border'));
    expect($sc.hasClass('success-form-border'));
  });

  it('"years" input responds to changes', () => {
    wrapper.setProps({ ...props, step: 1 });
    const selector =
      'select[data-id="How many years have you been a member of this bank?"]';
    const event = { target: { value: 1, name: 'yearsAtBank' } };

    wrapper.find(selector).simulate('change', event);
    expect(wrapper.find(selector).props().value).toEqual(1);
  });

  it('"months" input responds to changes', () => {
    const selector =
      'select[data-id="How many months have you been a member of this bank?"]';
    const event = { target: { value: '2', name: 'monthAtBank' } };

    wrapper.find(selector).simulate('change', event);
    expect(wrapper.find(selector).props().value).toBe('2');
  });

  it('onSubmit could be called', () => {
    const event = { preventDefault: jest.fn(), stopPropagation: jest.fn() };

    wrapper.find(SUBMIT).simulate('click', event);
    expect(wrapper.props().onSubmit).toBeCalled();
  });
});
