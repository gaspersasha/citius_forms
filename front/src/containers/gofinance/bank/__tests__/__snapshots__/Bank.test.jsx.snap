// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Bank /> renders without errors 1`] = `
<Bank
  bankDetails={
    Object {
      "accountNumber": "",
      "monthAtBank": "0",
      "sortCode": "",
      "yearsAtBank": "0",
    }
  }
  isDataSending={false}
  isLoading={false}
  onSubmit={[MockFunction]}
  step={0}
>
  <div
    className="formSectionWrapper"
  >
    <p
      className="formPreamble bankPreamble"
      data-bind="bankDetailform"
    >
      Some 
      <b>
        lenders require basic bank details
      </b>
      . Filling in the details below will get you a wider range of quotes
    </p>
    <AccountInfoStep
      accountNumber={
        Object {
          "status": "default",
          "step": 0,
          "validLength": 8,
          "validationType": "accountNumber",
          "value": "",
        }
      }
      errors={Object {}}
      onBlur={[Function]}
      onChange={[Function]}
      removeErrorForm={[Function]}
      sortCode={
        Object {
          "status": "default",
          "step": 0,
          "validLength": 8,
          "validationType": "sortCode",
          "value": "",
        }
      }
    >
      <div
        className="accountFormWrapper"
      >
        <div
          className="numberInputWrapper"
        >
          <InputNumber
            className="inputNumber"
            label="What is your account number?"
            name="accountNumber"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyPress={null}
            placeholder=""
            required={false}
            status="default"
            subTitle="Your bank account number is 8 digits long and can usually be found on your debit card, cheque book or online banking service"
            value=""
          >
            <Label
              customClasses=""
              name="accountNumber"
              required={false}
              status="default"
              title="What is your account number?"
              withSub={true}
            >
              <div
                className="label"
              >
                <label
                  data-id={false}
                  htmlFor="accountNumber"
                >
                  <span
                    className="title halfMargin"
                    data-id="label-title"
                  >
                    What is your account number?
                  </span>
                  <span
                    className="label-sub-title"
                  >
                    Your bank account number is 8 digits long and can usually be found on your debit card, cheque book or online banking service
                  </span>
                  <input
                    className="inputNumber"
                    name="accountNumber"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyPress={[Function]}
                    placeholder="What is your account number?"
                    type="number"
                    value=""
                  />
                </label>
              </div>
            </Label>
          </InputNumber>
        </div>
        <div>
          <InputMaskText
            className="inputMaskText"
            label="What is your Sort Code?"
            mask={
              Array [
                /\\\\d/,
                /\\\\d/,
                "-",
                /\\\\d/,
                /\\\\d/,
                "-",
                /\\\\d/,
                /\\\\d/,
              ]
            }
            name="sortCode"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Sort Code"
            required={false}
            status="default"
            subTitle="Your bank sort code is 6 digits long and can usually be found next to your account number"
            type="tel"
            value=""
          >
            <Label
              customClasses=""
              name="sortCode"
              required={false}
              status="default"
              title="What is your Sort Code?"
              withSub={true}
            >
              <div
                className="label"
              >
                <label
                  data-id={false}
                  htmlFor="sortCode"
                >
                  <span
                    className="title halfMargin"
                    data-id="label-title"
                  >
                    What is your Sort Code?
                  </span>
                  <span
                    className="label-sub-title"
                  >
                    Your bank sort code is 6 digits long and can usually be found next to your account number
                  </span>
                  <t
                    className="inputMaskText"
                    guide={false}
                    mask={
                      Array [
                        /\\\\d/,
                        /\\\\d/,
                        "-",
                        /\\\\d/,
                        /\\\\d/,
                        "-",
                        /\\\\d/,
                        /\\\\d/,
                      ]
                    }
                    name="sortCode"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder="Sort Code"
                    render={[Function]}
                    type="tel"
                    value=""
                  >
                    <input
                      className="inputMaskText"
                      defaultValue=""
                      name="sortCode"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      placeholder="Sort Code"
                      type="tel"
                    />
                  </t>
                </label>
              </div>
            </Label>
          </InputMaskText>
        </div>
      </div>
    </AccountInfoStep>
    <Button
      className="confirm"
      disabled={false}
      href=""
      id="ssFinanceBankDetails1Submit"
      onClick={[Function]}
      selected={false}
      styleType="primary"
      type="submit"
    >
      <button
        className="buttonComponent primary confirm"
        disabled={false}
        href={null}
        id="ssFinanceBankDetails1Submit"
        onClick={[Function]}
        type="submit"
      >
        Next
      </button>
    </Button>
  </div>
</Bank>
`;
